### ISSUE TITLE
Workaround for hidden twitch vods + Question

### ISSUE THREAD

**Issue by u2716** [DmitryScaletta/twitch-dlp#1]
Sometimes ongoing streams don't have a vod in the past broadcasts section, which causes twitch-dlp to fail the download of the stream from the start. Yet the link to the m3u8 vod can still be retrieved through [Shishkebaboo/VodRecovery](https://github.com/Shishkebaboo/VodRecovery) or [TwitchRecover/TwitchRecover](https://github.com/TwitchRecover/TwitchRecover).
Like a normal vod it is updated with new fragments every 5 minutes or so. It would be neat if twitch-dlp could also support these m3u8 links!

I have also a question. When yt-dlp downloads a livestream or vod from twitch it does a [FixupM3u8] by default. Does twitch-dlp also do a Fixup after merging the fragments?

**Comment by DmitryScaletta** [comment]
> Yet the link to the m3u8 vod can still be retrieved through [Shishkebaboo/VodRecovery](https://github.com/Shishkebaboo/VodRecovery) or [TwitchRecover/TwitchRecover](https://github.com/TwitchRecover/TwitchRecover)

Thanks for the suggestion. I tried to implement it in `v0.2.0`.
Didn't test it very well but it should work.

> Does twitch-dlp also do a Fixup after merging the fragments?

`yt-dlp` merges all raw fragments into one intermediate file to prevent storing multiple fragments in one folder and then does a [FixupM3u8].
`twitch-dlp` just downloads all fragments and then merges them into a result file with this `ffmpeg` command:

```bash
ffmpeg -n -f concat -safe 0 -i %fragsListFilename% -c copy %outputFilename%
```

So Fixup is not needed here.

**Comment by u2716** [comment]
Tested right now on a stream from amazonmusic which had the vod not public yet and seems working fine. 
The vod retrieval method works also for sub-only vods btw. Currently twitch-dlp doesn't try to recover the vod if the vod shows up and is sub-only, but with a few fixes it should work!

**Comment by DmitryScaletta** [comment]
> Currently twitch-dlp doesn't try to recover the vod if the vod shows up and is sub-only, but with a few fixes it should work!

Yeah, I'll do it later. Thanks for the information.

**Comment by DmitryScaletta** [comment]
> The vod retrieval method works also for sub-only vods btw.

Looks like it doesn't work right now from my testing.
Or they're using different base url that is not listed here:

https://github.com/DmitryScaletta/twitch-dlp/blob/master/twitch-dlp.js#L14-L30

**Comment by u2716** [comment]
I don't know. It seems to work for me. For instance, I search 'sub-only' on twitch and this is the first result I get `https://www.twitch.tv/onlykiky`. It's currently live and the vod is sub-only. Then I go to streamcharts for getting the stream id and start time `https://streamscharts.com/channels/onlykiky/streams/40654758597` . I run the python script `python.py vod_recovery.py` and successfully retrieve the VOD url `https://d1ymi26ma8va5x.cloudfront.net/2afb82bdd0da10f29285_onlykiky_40654758597_1715874746/chunked/index-dvr.m3u8`

**Comment by DmitryScaletta** [comment]
> Then I go to streamscharts for getting the stream id and start time

I can't use any third party services because usually they have some anti crawler protection.
I use twitch API: `VideoMetadata` gql request.

`createdAt` field from twitch api:

```ts
new Date('2024-05-16T15:52:32Z').getTime() / 1000; // 1715874752
```
Start time from streamcharts:

```ts
new Date('16 May 2024, 15:52Z').getTime() / 1000; // 1715874720
```

But in your example it `1715874746`.

I don't know why it works because it shouldn't.
The date doesn't even match the pattern:
https://github.com/Shishkebaboo/VodRecovery/blob/main/vod_recovery.py#L490

**Comment by u2716** [comment]
Streamscharts html data says that it started at `2024-05-16 15:52:26` so `1715874746` . I don't know why twitch API gives a different time...
I wonder what Streamscharts/Twitchtracker/etc use to scrape data then

**Comment by DmitryScaletta** [comment]
> So what is preventing to recover the URL of sub-only vods the same way twitch-dlp does for hidden vods? I mean, if you can retrieve the correct start time from the ongoing stream it should be feasible, right?

Yeah, it's possible to get the URL for sub-only ongoing streams, but not for sub-only VODs.
I fixed it in `0.2.2`.

> How does twitch-dlp handle muted fragments? yt-dlp retries 10 times on HTTP Error 403 before skipping to the next fragment. Does twitch-dlp do something similar?

Currently twitch-dlp doesn't have any functionality to handle muted fragments.
As far as I know it's possible to unmute muted fragments in first 24 hours after the stream.

**Comment by u2716** [comment]
>Yeah, it's possible to get the URL for sub-only ongoing streams, but not for sub-only VODs.
I fixed it in 0.2.2.
My bad, I didn't update.

>Currently twitch-dlp doesn't have any functionality to handle muted fragments.
As far as I know it's possible to unmute muted fragments in first 24 hours after the stream.
So let's say there are some muted fragments for copyright, they are skipped by yt-dlp because of the HTTP error 403. what happens in this case for twitch-dlp? Sorry if it's a stupid question. Just want to be sure.

I have some other enhancement requests or possible ideas. Do you want me to create one issue for each one of them or asking here is fine?
